src/OrbitBase/include/OrbitBase/ThreadPool.h

```diff
@@ -38,8 +40,28 @@ class ThreadPool {
   virtual void Schedule(std::unique_ptr<Action> action) = 0;
 
   template <typename F>
-  void Schedule(F&& functor) {
-    Schedule(CreateAction(std::forward<F>(functor)));
+  auto Schedule(F&& functor) -> orbit_base::Future<std::decay_t<decltype(functor())>> {
+    using ReturnType = std::decay_t<decltype(functor())>;
+
+    orbit_base::Promise<ReturnType> promise;
+    orbit_base::Future<ReturnType> future = promise.GetFuture();
+
+    if constexpr (std::is_same_v<ReturnType, void>) {
+      auto function_wrapper = [functor = std::forward<F>(functor),
+                               promise = std::move(promise)]() mutable {
```

Comment:
why is it mutable?

Reply:
It mutates `promise` by calling `promise.MarkFinished()`.

Reply:
Hmm. The more I think about it. Maybe `MarkFinished` should be `const`.

Reply:
Circling back to this: Making `MarkFinished` const is probably also not a good idea. I guess I prefer to leave the lambda `mutable`.